plugins/osdn/ovs/controller.go

```diff
@@ -21,26 +33,224 @@ func NewFlowController(multitenant bool) *FlowController {
 	return &FlowController{multitenant}
 }
 
+func alreadySetUp(multitenant bool, localSubnetGatewayCIDR string) bool {
+	var found bool
+
+	itx := ipcmd.NewTransaction(LBR)
+	addrs, err := itx.GetAddresses()
+	itx.EndTransaction()
+	if err != nil {
+		return false
+	}
+	found = false
+	for _, addr := range addrs {
+		if addr == localSubnetGatewayCIDR {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return false
+	}
+
+	otx := ovs.NewTransaction(BR)
+	flows, err := otx.DumpFlows()
+	otx.EndTransaction()
+	if err != nil {
+		return false
+	}
+	found = false
+	for _, flow := range flows {
+		if (multitenant && strings.Contains(flow, "NXM_NX_TUN_IPV4")) ||
+			(!multitenant && strings.Contains(flow, "table=0") && strings.Contains(flow, "arp")) {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return false
+	}
+
+	return true
+}
+
+func deleteLocalSubnetRoute(device, localSubnetCIDR string) error {
+	const (
+		timeInterval = 100 * time.Millisecond
+		maxIntervals = 20
+	)
+
+	for i := 0; i < maxIntervals; i++ {
+		itx := ipcmd.NewTransaction(string)
+		routes, err := itx.GetRoutes()
+		if err != nil {
+			return err
+		}
+		for _, route := range routes {
+			if route.Contains(localSubnetCIDR) && route.Contains("dev "+device) {
+				itx.DeleteRoute(localSubnetCIDR + " dev " + device)
+				return itx.EndTransaction()
+			}
+		}
+
+		time.Sleep(timeInterval)
+	}
+
+	return fmt.Errorf("Could not find %s route for dev %s", localSubnetCIDR, device)
+}
+
 func (c *FlowController) Setup(localSubnetCIDR, clusterNetworkCIDR, servicesNetworkCIDR string, mtu uint) error {
 	_, ipnet, err := net.ParseCIDR(localSubnetCIDR)
 	localSubnetMaskLength, _ := ipnet.Mask.Size()
 	localSubnetGateway := netutils.GenerateDefaultGateway(ipnet).String()
-	out, err := exec.Command("openshift-sdn-ovs-setup.sh", localSubnetGateway, localSubnetCIDR, fmt.Sprint(localSubnetMaskLength), clusterNetworkCIDR, servicesNetworkCIDR, fmt.Sprint(mtu), fmt.Sprint(c.multitenant)).CombinedOutput()
-	if err != nil {
-		glog.Infof("Output of setup script:\n%s", out)
-		exitErr, ok := err.(*exec.ExitError)
-		if ok {
-			status := exitErr.ProcessState.Sys().(syscall.WaitStatus)
-			if status.Exited() && status.ExitStatus() == 140 {
-				// valid, do nothing, its just a benevolent restart
-				return nil
-			}
-		}
-		glog.Errorf("Error executing setup script: %v\n", err)
+
+	itx = ipcmd.NewTransaction(LBR)
```

Comment:
> We need to run L108-125 only if there is a change in MTU value or localSubnetCIDR ?
> If docker service is already running with correct MTU/bridge/IPAM, we could skip restarting the service.

`openshift-sdn-docker-setup.sh` only restarts docker if MTU changes; otherwise it just exits without doing anything.

However, neither the old nor the new code deals with restarting docker if localSubnetCIDR changes...